authenticate.js

```diff
@@ -1,41 +1,57 @@
 const alfy = require('alfy');
 const bitbucket = require('./bitbucket/core').bitbucket;
 
-const { clientId, secret } = process.env;
-
 const ACCESS_TOKEN = 'access_token';
 const GRANT_TYPE = 'client_credentials';
 const ALIVE_TIME = 216000;
 const URL = 'https://bitbucket.org/site/oauth2/access_token';
 
-const OPTIONS = {
-    auth: [clientId, secret].join(':'),
-    json: true,
-    method: 'POST',
-    body: {
-        grant_type: GRANT_TYPE
-    },
-    form: true
-};
-
-const loginAttempt = () => {
-    return new Promise((resolve, _reject) => {
-        if (!clientId || !secret) {
-            alfy.error('OAuth2 not set. Refer to README for details');
-        } else {
-            const accessToken = alfy.cache.get(ACCESS_TOKEN);
-            if (accessToken) {
-                resolve(bitbucket(accessToken));
-            } else {
-                alfy.fetch(URL, OPTIONS).then(({ access_token }) => {
-                    alfy.cache.set(ACCESS_TOKEN, access_token, { maxAge: ALIVE_TIME });
-                    resolve(bitbucket(access_token));
-                }).catch(() => {
-                    alfy.cache.set(ACCESS_TOKEN, null);
-                });
-            }
-        }
-    });
-};
+function getAppPasswordToken(username, password) {
+    return `Basic ${Buffer.from([username, password].join(':')).toString('base64')}`;
+}
+
+async function getAuthToken(clientId, secret) {
+    const accessToken = alfy.cache.get(ACCESS_TOKEN);
+
+    if (accessToken) {
+        return bitbucket(accessToken);
+    }
+
+    try {
+        const {access_token} = await alfy.fetch(URL, {
+            auth: [clientId, secret].join(':'),
+            json: true,
+            method: 'POST',
+            body: {
+                grant_type: GRANT_TYPE
+            },
+            form: true
+        });
+        const authToken = `Bearer ${access_token}`;
+
+        alfy.cache.set(ACCESS_TOKEN, authToken, {maxAge: ALIVE_TIME});
+
+        return authToken;
+    } catch (e) {
+        alfy.cache.set(ACCESS_TOKEN, null);
+    }
+
+    return null;
+}
+
+async function loginAttempt() {
+    const {clientId, secret, appPassword, username} = process.env;
+
+    if ((!username || !appPassword) && (!clientId || !secret)) {
+        alfy.error('OAuth2 not set. Refer to README for details');
+        return null;
+    }
+
+    if (username && appPassword) {
```

Comment:
The goal of my comment was to remove as much as possible these conditions in a way the check are performed once and, if possible, only one invocation of `bitbucket()`. What you have there is really nice. I would just still split it so the token is obtained in one function and another on calls it and if it's null it shows your error like
```js
const getToken = async () => {
  if (username && appPassword) {
    return getAppPasswordToken(username, appPassword)
  }
  if (clientId && secret) {
    return await getAuthToken(clientId, secret)
  }
  
  return null
}

async function loginAttempt() {
  const token = getToken()

  if (token === null) {
    return alfy.error('OAuth2 not set. Refer to README for details');
  }

  return bitbucket(token)
}
``` 
What you have here with this addition would be perfect for me 👍

Reply:
Ahh okay, yeah I missed what you were asking for. I agree, although I don't believe there needs to be a `getToken` function. Lemme refactor some more and see if you like it